Expose SIMD UTF-8 validation functions from internal module

andrewthad opened the pull request:
This makes it possible for users to validate ByteArray# when compiling with the SIMDUTF8 flag on. It does not, however, provide any fallbacks for when that flag is off. That is left as future work.

Related to https://github.com/haskell/text/issues/479

andrewthad commented:
I've added portable functions that work without SIMDUTF being turned on. This is ready for review.

andrewthad commented:
Fixed a problem with GHC 8.0. I still need to add some tests to the test suite. The `ByteArray#` functions have no coverage.

andrewthad commented:
Just squashed this down to remove a bunch of meaningless commits.

Bodigrim commented on src/Data/Text/Internal/PrimCompat.hs:
```diff
@@ -10,16 +10,23 @@ module Data.Text.Internal.PrimCompat
 import GHC.Prim (Word#)
+#endif
 
+#if MIN_VERSION_base(4,10,0)
```
I think it's time for `base-4.9` (and corresponding GHC 8.0) to go: I'd rather drop its support than accumulate more boilerplate. @Lysxia what do you think?

Lysxia replied:
I also think we can drop base-4.9.

andrewthad replied:
I’m fine with killing off GHC 8.0 support. Should I open a separate PR doing that? I think it would be best for that to be a separate commit and not tied up in this addition.

Bodigrim replied:
Yes, please make a separate PR to decomission GHC 8.0.

andrewthad replied:
Done.

andrewthad commented:
This is ready for another round of review now that support for GHC 8.0 is gone. The only changes since last time are documentation. I've provided the motivation for the module in the module's haddock, and I've documented the behaviors of the functions.

Bodigrim commented on src/Data/Text/Internal/Validate.hs:
```diff
@@ -0,0 +1,155 @@
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE UnliftedFFITypes #-}
+#if defined(SIMDUTF) || MIN_VERSION_bytestring(0,11,2)
+{-# LANGUAGE Trustworthy #-}
```
`Data.Text.Encoding` used to promise `Trustworthy` unconditionally. Any particular reason not to follow its example?

I would rather remove `{-# LANGUAGE Trustworthy #-}` (not a fan of "Safe" Haskell) from both modules than promise it only conditionally, because it is likely to explode somewhere later in the ecosystem.

andrewthad replied:
I'll just remove it. I'm not even sure what guarantees Safe Haskell is supposed to provide, but it's trivial to trigger a segfault by misusing the functions exposed by this module.

Bodigrim commented on src/Data/Text/Internal/Validate.hs:
```diff
@@ -0,0 +1,155 @@
+import Data.Text.Internal.Unsafe (unsafeWithForeignPtr)
+import Data.Text.Internal.Validate.Simd (c_is_valid_utf8_bytearray_safe)
+import Data.Text.Internal.Validate.Simd (c_is_valid_utf8_bytearray_unsafe)
+import Data.Text.Internal.Validate.Simd (c_is_valid_utf8_ptr_unsafe)
```
Please collapse these three imports.

andrewthad replied:
Will do.

Bodigrim commented on src/Data/Text/Internal/Validate.hs:
```diff
@@ -0,0 +1,155 @@
+import Data.Text.Internal.Validate.Simd (c_is_valid_utf8_ptr_unsafe)
+#else
+import Data.Text.Internal.Encoding.Utf8 (CodePoint(..),DecoderResult(..))
+import Data.Text.Internal.Encoding.Utf8 (utf8DecodeStart,utf8DecodeContinue)
```
And these two.

andrewthad replied:
Will do.

Bodigrim commented on src/Data/Text/Internal/Validate.hs:
```diff
@@ -0,0 +1,155 @@
+import qualified Data.ByteString as B
+#else
+import qualified Data.ByteString as B
+import qualified Data.ByteString.Unsafe as B
```
Let's `import qualified Data.ByteString as B` unconditionally.

Bodigrim commented on text.cabal:
```diff
@@ -170,6 +170,8 @@ library
     Data.Text.Internal.Unsafe
     Data.Text.Internal.Unsafe.Char
+    Data.Text.Internal.Validate
+    Data.Text.Internal.Validate.Simd
```
Please make this module conditional: `if flag(simdutf) exposed-modules: Data.Text.Internal.Validate.Simd`

andrewthad replied:
Will do. Also, I'll remove the conditionals from the `Simd` module itself since it's compilation will be guarded in this way.

Bodigrim commented on src/Data/Text/Internal/Validate.hs:
```diff
@@ -0,0 +1,147 @@
+  = isValidUtf8ByteArraySafe# b off len
+  | otherwise = isValidUtf8ByteArrayUnsafe# b off len
+
+-- | This always uses the unsafe FFI. It is always safe to call this
```
I understand what you mean, but I imagine plenty of people puzzled why it is always safe to call unsafe FFI. Could you please elaborate a little bit more, "the unsafe FFI, which is guaranteed not to be interrupted by GC" or similar?

andrewthad replied:
I've elaborated more in latest commit.

Bodigrim commented on src/Data/Text/Internal/Validate.hs:
```diff
@@ -0,0 +1,147 @@
+
+-- | This always uses the safe FFI. The byte array /must/ be pinned,
+-- and the calling context is responsible for enforcing this. If the
+-- byte array is not pinned, this function's behavior is undefined.
```
Same as above, let's elaborate for less experienced users about the interaction between safe FFI, pinnedness and GC.

Bodigrim commented on src/Data/Text/Internal/Validate.hs:
```diff
@@ -0,0 +1,147 @@
+-- Note: The release of GHC 9.4 has seen a coordinated move of the lifted
+-- @ByteArray@ type from @primitive@ to @base@. In many years, when support
+-- for GHC older than 9.4 is dropped, it will be possible to add functions
+-- for the lifted ByteArray type as well.
```
Please rebase your branch atop of https://github.com/haskell/text/pull/474 and switch types to a lifted `type Array = ByteArray` from `Data.Text.Array`. I'm a bit reluctant to introduce API for unlifted arrays just to be painfully deprecated (or extended) soon.

Bodigrim replied:
I have no strong preferences here.

andrewthad commented:
Failure on GHC 8.6.5 looks spurious.

andrewthad commented:
Just squashed. This will kick off another round of CI.

andrewthad commented:
CI is still failing, but I’m pretty sure that it’s spurious. This is ready for another round of review

andrewthad commented:
Wait, something really is wrong on the ppcle64 architecture. For some reason, the dependency on the byte array compatibility library doesn’t get picked up, and so it ends up missing the module. I don’t see anything wrong in the Cabal file though.

Bodigrim commented:
@andrewthad necessary changes are in https://github.com/haskell/text/pull/474/commits/d0b70f782b22ca3d39c35b3cbc042757340cf8b1. I think you squashed a bit overzealously.

andrewthad commented:
I’ve incorporated the diff manually. Thanks. Looks like that fixed it.

andrewthad commented:
Only remaining CI failure seems unrelated to the change.

Bodigrim commented:
It's good to go, but on hold waiting for #474 to land (which in its turn waits until we cut a release to accompany GHC 9.6).

Bodigrim commented:
@Lysxia given that #448 has landed, could you take a look at what remains of this PR? I assume the new design is flexible enough already, but we might still want to expose low-level calls to simdutf?..

I’m AFK and cannot take a close look myself.

andrewthad commented:
Is there anything I can do to move this forward? Should I resolve the merge conflict, or are there concerns about this feature no longer fitting into the design of `text`?

Lysxia commented:
We can still accept this. The bytearray validation functions are new. And while we now have a fancier bytestring validation function, adding the bool version would look consistent with the bytearray ones.

Lysxia commented on src/Data/Text/Internal/Validate.hs:
```diff
@@ -0,0 +1,149 @@
+-- @safe@ FFI call is executing. The byte array argument /must/ be pinned,
+-- and the calling context is responsible for enforcing this. If the
+-- byte array is not pinned, this function's behavior is undefined.
+isValidUtf8ByteArraySafe ::
```
Instead of `Unsafe/Safe` why not `Unpinned/Pinned`?

The `Safe/Unsafe` naming is confusing since "safe FFI" calls are unsafe and "unsafe FFI" calls are safe. The FFI functions themselves don't need to be renamed.

We could also consider `Unsafe/Pinned` but this still gives the wrong impression that the pinned one is safer than `Unsafe`.

Lysxia commented on src/Data/Text/Internal/Validate.hs:
```diff
@@ -0,0 +1,149 @@
+-- run concurrently with GC and is not interrupted by GC. Since relocation
+-- cannot happen concurrently with an @unsafe@ FFI call, it is safe
+-- to call this function with an unpinned byte array argument.
+-- (It is also safe to call this with a pinned @ByteArray@ argument.)
```
```suggestion
-- It is also safe to call this with a pinned @ByteArray@ argument.
```

That's really a proper part of the spec; it's even already relied upon by the other function.

andrewthad commented on src/Data/Text/Internal/Validate.hs:
```diff
@@ -0,0 +1,149 @@
+#endif
+#endif
+
+-- | For pinned byte arrays larger than 128KiB, this switches to the safe FFI
```
I'm going to answer this on the main thread, and we can discuss it there. I started writing this, and I realized there's enough information that I don't want github's comment folding to make it impossible for future readers to find it.

Lysxia commented:
That's a very clear analysis! That makes sense to me. Confirming this further, the GHC User Manual also says:

> This means that if you need to make a foreign call to a function that takes a long time or potentially blocks, then you should mark it safe and use -threaded. Some library functions make such calls internally; their documentation should indicate when this is the case.
>
> On the other hand, a foreign call to a function that is guaranteed to take a short time, and does not call back into Haskell can be marked unsafe.

Bodigrim commented:
@andrewthad could you please rebase?

andrewthad commented:
I've just now squashed everything down and rebased.

Bodigrim commented on src/Data/Text/Internal/Validate.hs:
```diff
@@ -0,0 +1,149 @@
+  ) where
+
+import Data.Array.Byte (ByteArray(ByteArray))
+import GHC.Exts (ByteArray#)
```
Redundant import at least for recent GHCs.

andrewthad replied:
I've now guarded this behind CPP checking that simdutf is not defined.

Bodigrim commented on text.cabal:
```diff
@@ -136,6 +137,10 @@ library
 
+  -- Before GHC 9.4, Data.Array.Byte is not in base
+  if impl(ghc < 9.4)
+    build-depends: data-array-byte >= 0.1.0.1
```
This is redundant, `build-depends: data-array-byte` is already in `text.cabal`.

Bodigrim commented on text.cabal:
```diff
@@ -85,6 +85,7 @@ library
   hs-source-dirs: src
 
   if flag(simdutf)
+    exposed-modules: Data.Text.Internal.Validate.Simd
```
Please mention this additional module in the flag description above.

Bodigrim commented on src/Data/Text/Internal/Validate/Simd.hs:
```diff
@@ -0,0 +1,43 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE CPP #-}
```
Seems there is no CPP used?

andrewthad commented:
I'm not able to make sense of the NetBSD test failure output.

andrewthad commented:
Is there anything else that I can improve in this PR, or is this ready to be merged?
